File: 20-ValidParentheses/20-ValidParentheses.java
/*

Given a string containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

An input string is valid if:

    Open brackets must be closed by the same type of brackets.
    Open brackets must be closed in the correct order.

Note that an empty string is also considered valid.

Example 1:

Input: "()"
Output: true

Example 2:

Input: "()[]{}"
Output: true

Example 3:

Input: "(]"
Output: false

Example 4:

Input: "([)]"
Output: false

Example 5:

Input: "{[]}"
Output: true

Question Link : https://leetcode.com/problems/valid-parentheses/

*/

class Solution {
    public boolean isValid(String s) {
        if (s.length()==0)
            return true;
        if (s.length()==1)
            return false;
        
        Stack <Character> tempStack = new Stack<>();
        Map <Character,Character> lookupMap = new HashMap<>();
        lookupMap.put('(',')');
        lookupMap.put('[',']');
        lookupMap.put('{','}');
        for (int i=0;i<s.length();i++)
        {
            if(lookupMap.containsKey(s.charAt(i)))
                tempStack.push(lookupMap.get(s.charAt(i)));
            else
            {
                if(tempStack.isEmpty()||s.charAt(i)!=tempStack.pop())
                    return false;
            }
        }
        return tempStack.isEmpty();
    }
}

File: 20-ValidParentheses/20-ValidParentheses.py
"""

20. Valid Parentheses

Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

An input string is valid if:

    Open brackets must be closed by the same type of brackets.
    Open brackets must be closed in the correct order.
    Every close bracket has a corresponding open bracket of the same type.

Example 1:

Input: s = "()"
Output: true

Example 2:

Input: s = "()[]{}"
Output: true

Example 3:

Input: s = "(]"
Output: false

Constraints:

    1 <= s.length <= 104
    s consists of parentheses only '()[]{}'.

"""


class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        mapper = {"(" : ")",
                  "[" : "]",
                  "{" : "}"};

        for i in range(len(s)):
            ch = s[i]
            if len(stack) == 0:
                if ch not in mapper.keys():
                    return False
                else:
                    stack.append(mapper[ch])
            else:
                if ch not in mapper.keys():
                    if stack.pop() != ch:
                        return False
                else:
                    stack.append(mapper[ch])

        return len(stack) is 0
        
